### PR TITLE
feat(listener): let a first message from a never-tracked chat match CHAT_TYPES

### LINKED ISSUE #415: Listener misses first message from a chat we've never backed up (new DMs/groups invisible until next scheduled run)
## Bug description

The real-time listener's `_should_process_chat()` (src/listener.py) is intentionally conservative for chats it has never seen before: since it only receives a bare `chat_id`, it can't cheaply determine the chat's type, so a chat that is neither already in `_tracked_chat_ids` (i.e. already backed up at least once) nor in an explicit `*_INCLUDE_CHAT_IDS` list is dropped, regardless of `CHAT_TYPES`.

In practice this means: a brand-new DM from someone you've never chatted with before (or a new group/channel you just joined) is invisible to `ENABLE_LISTENER=true` until the next *scheduled* backup run discovers it via `get_dialogs()` and adds it to the database — which, depending on `SCHEDULE`, can be hours later. Only messages *after* that point are captured live.

## Steps to reproduce

1. Run with `ENABLE_LISTENER=true`, `CHAT_TYPES=private` (no `PRIVATE_INCLUDE_CHAT_IDS` set), and a `SCHEDULE` of several hours.
2. Have someone who has never messaged you before send you a Telegram DM.
3. The message does not appear in the archive/viewer until the next scheduled backup runs — even though the listener is connected and receiving the `NewMessage` update the whole time.

## Expected behavior

Since `events.NewMessage.Event` already carries `is_private`/`is_group`/`is_channel` synchronously (derived from the update's own peer, no entity fetch/network round-trip required), the listener could evaluate a never-before-seen chat against `CHAT_TYPES` immediately instead of only against the include lists, and capture first contact live.

## Proposed fix

I have a working patch + tests for this (extends `_should_process_chat()` with optional `is_user`/`is_group`/`is_channel` hints, used by `on_new_message`, falling back to `Config.should_backup_chat_type()` for untracked chats) and will open a PR.

### PR DESCRIPTION
**Pull request by jordanfelle**
## Summary

- Fixes #415: `ENABLE_LISTENER=true` currently drops the very first message from a chat you've never backed up before (a new DM from someone you've never chatted with, or a group/channel you just joined), because `_should_process_chat()` can't cheaply tell a bare `chat_id`'s type and conservatively only matches already-tracked chats or explicit `*_INCLUDE_CHAT_IDS` entries. First contact only appears once the next *scheduled* backup discovers the chat via `get_dialogs()`.
- `events.NewMessage.Event` already exposes `is_private`/`is_group`/`is_channel` synchronously (from the update's own peer, no entity fetch/network round-trip), so `on_new_message` now passes those through as optional hints to `_should_process_chat()`, which falls back to `Config.should_backup_chat_type()` for a chat that's neither tracked nor in an include list.
- Scope: only `on_new_message` supplies the hint. Every other handler (edits, deletions, pins, reactions, chat actions) is untouched — those events only ever matter for a chat a `NewMessage` has already made known, so they can't hit the "never seen before" case this PR addresses.
- Whitelist mode (`CHAT_IDS` set) is unaffected — it's checked first and returns before the hint is ever consulted.
- Known limitation (documented in the docstring): a first DM from a bot is evaluated against CHAT_TYPES' `"private"` bucket rather than `"bots"`, since telling the two apart needs the sender entity, which isn't available synchronously off the event. The next scheduled backup reclassifies the chat correctly either way, so this only affects the very first live-captured message in that edge case.

## Test plan

- [x] `ruff check .` / `ruff format --check .` clean
- [x] `python -m pytest tests/ -v` — 3765 passed, 138 skipped (same skip count as `main`), 0 failed
- [x] Added unit tests for `_should_process_chat`'s new hint kwargs (no-hint stays conservative, hint match/no-match, whitelist mode ignores the hint) in `tests/test_listener.py`
- [x] Added `on_new_message` end-to-end tests for a genuinely first-contact chat: one where the CHAT_TYPES hint matches (message captured, chat tracked) and one where it doesn't (message dropped, chat stays untracked)
- [x] Fixed two other test fixtures/builders (`test_listener.py`'s `mock_config`/`full_config`, `test_listener_reactions.py`'s `_new_message_event`) that previously left `should_backup_chat_type`/`is_private`/`is_group`/`is_channel` unset — those now default to the pre-PR conservative behavior explicitly rather than relying on `MagicMock()`'s default truthiness

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved scheduled-backup decisions for new and untracked chats, including exclusions and include-list rules.
  - Chat ID allowlists now correctly work alongside chat-type settings.
  - Megagroups are classified as groups, while chats with unknown channel status are handled conservatively.
  - First-contact messages are processed according to the complete backup configuration.

- **Tests**
  - Expanded coverage for backup precedence, chat classification, allowlists, and ambiguous channel events.

### CODE REVIEW

**Comment by jordanfelle** [comment]
Addressed the CodeRabbit finding in 540a12f: \`event.is_channel\` is true for any \`PeerChannel\` (megagroup or broadcast channel) while \`event.is_group\` is also true for a megagroup, so both flags were true simultaneously and a megagroup's first message could have matched a channels-only \`CHAT_TYPES\` filter it should've been excluded from - \`_get_chat_type()\` elsewhere in this file already treats a megagroup as `"group"`, never `"channel"`. Now clearing the `is_channel` hint whenever `is_group` is also set, with a regression test (`test_on_new_message_first_contact_megagroup_not_treated_as_channel`) that fails without the fix and passes with it.

Left the test-annotation suggestion (`-> None` on the new test methods) as-is - no other test in this file uses that annotation, so adding it only to the new ones would be inconsistent with the surrounding code.

**Comment by GeiserX** [comment]
Great fix, and the megagroup follow-up was exactly right — thank you. I pushed one commit on top (d91b331) tightening two things review turned up:

1. The fallback consulted `should_backup_chat_type()` directly, which skips the exclude lists and the "an include list acts as a whitelist" semantics — so a first message from a chat you'd excluded (or outside your include-whitelist) whose type matches `CHAT_TYPES` would have been live-captured and tracked, while the scheduled backup would refuse that same chat. It now delegates to `should_backup_chat()` — the exact decision the scheduled backup makes — so the two paths can't disagree.

2. Telethon's `ChatGetter.is_group` returns `None` for a `PeerChannel` when it can't see the broadcast flag (chat entity absent from the update). `None` is falsy, so the `is_channel and is_group` normalization silently did nothing in exactly the case it was written for, and an unknown megagroup would have been evaluated as "channel". When that happens we now pass no hints at all — that one message takes the conservative pre-hint path and the next scheduled backup picks the chat up, same as before this feature.

Tests updated accordingly, plus regression tests for both (`test_should_process_chat_type_hint_uses_full_decision_not_bare_type_filter`, `test_on_new_message_channel_unknown_broadcast_stays_conservative`).

File: tests/test_listener.py [hunk 1]
```diff
@@ -88,6 +91,42 @@ def test_should_process_chat_include_list(self, mock_config, mock_db):
         assert listener._should_process_chat(-1009999999) is True
         assert listener._should_process_chat(-1008888888) is False
 
+    def test_should_process_chat_no_hint_stays_conservative(self, mock_config, mock_db):
+        """Without a type hint, an untracked/unlisted chat is still rejected (#415 regression guard)."""
+        listener = TelegramListener(mock_config, mock_db, account_id=1)
+        listener._tracked_chat_ids = set()
+
+        assert listener._should_process_chat(555) is False
+        mock_config.should_backup_chat_type.assert_not_called()
```
[the remark this answers is not in the document]
Comment (GeiserX) [reply tests/test_listener.py]: Agreed with jordanfelle here — no other test in this file annotates `-> None`, and per-file consistency beats the annotation. Not adopting it.

**Comment by GeiserX** [comment]
Merged — thank you for this one. First contact being invisible until the next scheduled sweep was a real gap in the listener, your fix took the cheap synchronous route (no entity fetch on the hot path), and the megagroup follow-up with a regression test that fails without it is exactly how I like these to arrive. The test-fixture cleanup for the MagicMock truthiness trap was appreciated too.
